### PR TITLE
Send v2

### PR DESCRIPTION
**Pull request by ftreguer**
update send command to enable CLI users to send UNS tokens to unik-names [#1670](https://spacelephant.tpondemand.com/entity/1670-ron-je-veux-envoyer-des-uns)

- Update send command
- Some new exit case tests
- Extract functions from send.ts file, moved to commandHelpers/send-helper.ts
- Fix getVersion network function (impact set-properties and unset-properties command)

### CODE REVIEW

File: src/updatePropertiesCommand.ts [hunk 1]
```diff
@@ -12,13 +12,15 @@ import {
...
 
 export abstract class UpdateProperties extends WriteCommand {
-    public static flags = {
-        ...WriteCommand.flags,
-        ...unikidFlag("The UNIK token on which to update the properties."),
-        ...awaitFlag,
-        ...confirmationsFlag,
-        ...passphraseFlag,
-    };
+    protected static getUpdateCommandFlags(fees: number = 100000000) {
```
Comment (Nigui) [inline src/updatePropertiesCommand.ts]: I'm not convinced of default value duplicated in every functions. 
A function calling another with an optional parameter should not declare a default value. It should delegate optional to called function.

Comment (Nigui) [reply src/updatePropertiesCommand.ts]: and that default value should be a constant

File: test/__fixtures__/commands/send.ts [hunk 2]
```diff
@@ -112,4 +112,14 @@ export const shouldExit = [
         args: ["send", "--network", "devnet", "10", "--to", "DDwxZwjZQJUjeu7P"],
         exitCode: 1,
     },
+    {
+        description: "Should exit with code 1 recipient is not a valid @unik-name",
+        args: ["send", "--network", "devnet", "10", "--to", "@test?*"],
```
Comment (Nigui) [inline test/__fixtures__/commands/send.ts]: PO said that `@test?*` is valid

File: src/commands/send.ts [hunk 3]
```diff
@@ -69,175 +72,45 @@ export class SendCommand extends WriteCommand {
     }
 
     protected async do(flags: Record<string, any>, args?: Record<string, any>): Promise<any> {
+        if (flags.to === flags.senderAccount) {
```
Comment (Nigui) [inline src/commands/send.ts]: I think that control must be done later, when sender and recipient addresses are resolved.  
I've asked PO

File: src/commandHelpers/send-helper.ts [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        } catch (e) {
+            if (e && e.response && e.response.status === 404) {
+                throw new Error(`${isRecipient ? "Recipient" : "Sender"} @unik-name does not exist`);
+            }
+            throw new Error(`${isRecipient ? "Recipient" : "Sender"} @unik-name does not match expected format`);
+        }
+    } else {
+        if (!crypto.validateAddress(id, this.api.getVersion())) {
+            try {
+                resolvedAddress = Address.fromPublicKey(id);
```
Comment (Nigui) [inline src/commandHelpers/send-helper.ts]: it seems that we can pass a public key to `--to` flag. 
Maybe we could add it to documentation ? Or do we prefer to keep it as hidden feature ?

File: src/commandHelpers/send-helper.ts [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    checkPassphraseFormat,
+    createTransferTransaction,
+    getPassphraseFromUser,
+    getSecondPassphraseFromUser,
+    getWalletFromPassphrase,
+    toSatoshi,
+} from "../utils";
+
+export function getFees(isSato: boolean, fee: number): number {
+    return isSato ? (fee === SendCommand.DEFAULT_FEES ? toSatoshi(fee) : fee) : toSatoshi(fee);
```
Comment (Nigui) [inline src/commandHelpers/send-helper.ts]: Current function won't work if user pass `--fee 1 --sato`, if he wants to broadcast his transaction with fee=1.
you must refactor this function. 
If user doesn't pass `--fee` flag, it should be known by command process. Then, you can set default value and convert it to satoshi.

File: src/commandHelpers/send-helper.ts [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+            resolvedAddress = id;
+        }
+    }
+    return resolvedAddress;
+}
+
+export async function checkAndConfirmWallet(checkWalletExistence: boolean, address: string, passphrase?: string) {
+    if (checkWalletExistence) {
+        this.actionStart("Check recipient existence");
+        const exists = await this.checkWalletExistence(address);
```
Comment (Nigui) [inline src/commandHelpers/send-helper.ts]: Where is `checkWalletExistence` function ? 
Who is `this`?

File: src/commands/send.ts [hunk 7]
```diff
@@ -27,12 +24,14 @@ export class SendCommand extends WriteCommand {
...
-        ...WriteCommand.flags,
+        ...WriteCommand.getWriteCommandFlags(SendCommand.DEFAULT_FEES),
         ...passphraseFlag,
         ...secondPassphraseFlag,
         to: flags.string({
-            description: "The recipient address.",
+            description: "The recipient public address OR the @unik-name of the recipient.",
             required: true,
         }),
         check: flags.boolean({
```
Comment (Nigui) [inline src/commands/send.ts:35]: `check` description is not good. It says that `--check` "allow sending tokens to an address that do not exists on chain". 
It's the opposite, if you set `--check` it won't allow sending tokens to an address that does not exist on chain.

and spelling is wrong. It's "that does not exist on chain"

**Review by Nigui — CHANGES_REQUESTED** [review]
you must add successful test cases (exit code 0).
I tried : 
```
{
        description: "Should exit with code 0",
        args: ["send", "--network", "devnet", "10", "--to", "@test"],
        exitCode: 0,
    }
```
and it fails

**Comment by ftreguer** [comment]
> you must add successful test cases (exit code 0).
> I tried :
> 
> ```
> {
>         description: "Should exit with code 0",
>         args: ["send", "--network", "devnet", "10", "--to", "@test"],
>         exitCode: 0,
>     }
> ```
> 
> and it fails

You need a passphrase to send tokens

File: src/commandHelpers/send-helper.ts (hunk 4 above)
Comment (ftreguer) [reply src/commandHelpers/send-helper.ts]: TODO: update documentation

File: src/commandHelpers/send-helper.ts (hunk 6 above)
Comment (ftreguer) [reply src/commandHelpers/send-helper.ts]: Right, fail on refactoring, I'll create an helper with command instance

File: src/commandHelpers/send-helper.ts (hunk 4 above)
Comment (Nigui) [reply src/commandHelpers/send-helper.ts]: voluntarily missed ?

**Comment by Nigui** [comment]
> > you must add successful test cases (exit code 0).
> > I tried :
> > ```
> > {
> >         description: "Should exit with code 0",
> >         args: ["send", "--network", "devnet", "10", "--to", "@test"],
> >         exitCode: 0,
> >     }
> > ```
> > 
> > 
> > and it fails
> 
> You need a passphrase to send tokens

ok, but first comment is voluntarily missed ?

File: src/commandHelpers/send-helper.ts (hunk 4 above)
Comment (ftreguer) [reply src/commandHelpers/send-helper.ts]: Done in last commit

**Comment by ftreguer** [comment]
> > > you must add successful test cases (exit code 0).
> > > I tried :
> > > ```
> > > {
> > >         description: "Should exit with code 0",
> > >         args: ["send", "--network", "devnet", "10", "--to", "@test"],
> > >         exitCode: 0,
> > >     }
> > > ```
> > > 
> > > 
> > > and it fails
> > 
> > 
> > You need a passphrase to send tokens
> 
> ok, but first comment is voluntarily missed ?

No integration tests, just unit tests on specific functions and PO tests :)

**Review by Nigui — APPROVED** [review]

**Pull request merged**